File: sdk/go/cache.go
package chunkyourskills

// EnsureSkillsRegistry ensures page index (+ BM25 chunks when pipeline is bm25) for skill sources.
// sourcePathsJSON is a JSON array of filesystem paths (strings) or inline source objects
// with path, optional content, and optional content_sha256 (hook/client skills).
func EnsureSkillsRegistry(sourcePathsJSON, catalogRoot, pageindexConfigJSON, pipeline, indexParamsHash, policy string) (string, error) {
	return cgoEnsureSkillsRegistry(sourcePathsJSON, catalogRoot, pageindexConfigJSON, pipeline, indexParamsHash, policy)
}

// ConfigureMemoryCache applies in-memory cache tuning (lazy registry, LRU sizes, async disk writes).
func ConfigureMemoryCache(configJSON string) error {
	return cgoConfigureMemoryCache(configJSON)
}

File: sdk/typescript/src/cache.ts
/** Rust-backed disk/memory cache for skills registries. */

import {
  configureMemoryCacheNative,
  ensureSkillsRegistryNative,
} from "./native.js";

export type CachePolicy = "auto" | "force_memory" | "force_disk";

export interface SkillEntryRef {
  entry_dir: string;
  doc_id: string;
  content_sha256: string;
  bm25_chunk_dir: string | null;
  disk_backed: boolean;
  cache_status: "hit" | "miss" | "memory_fallback";
  source_path: string;
  nodes_dir: string | null;
  document: Record<string, unknown>;
  lazy_pending?: boolean;
}

/** Filesystem path or in-memory hook/client skill payload. */
export type SkillSourceInput =
  | string
  | {
      path: string;
      content?: string;
      content_sha256?: string;
    };

export function ensureSkillsRegistry(
  sourcePaths: SkillSourceInput[],
  catalogRoot: string,
  pageindexConfig: Record<string, unknown> | null | undefined,
  pipeline: string,
  indexParamsHash: string,
  policy?: CachePolicy,
): SkillEntryRef[] {
  return ensureSkillsRegistryNative(
    sourcePaths,
    catalogRoot,
    pageindexConfig ?? undefined,
    pipeline,
    indexParamsHash,
    policy,
  ) as SkillEntryRef[];
}

export function configureMemoryCache(config: Record<string, unknown>): void {
  configureMemoryCacheNative(config);
}
